### DISCUSSION TITLE
Cartridge freeze

### DISCUSSION THREAD

**Discussion by sgrockfort** [vossstef/tang_nano_20k_c64#93] [Ideas]
Hello @vossstef , thank you once more for adding the joystick swap shortcut. 
Would it also be possible to add a shortcut for activating the freeze button function of a cartridge? or by the F12 menu ? :-)

Thanks..
Michael

**Comment by vossstef** [comment]
Hi @sgrockfort ,
cartridge Freeze is invoked by F11 = RESTORE on C64 Keyboard and also NMI input of CPU.

I didn't tested freezer card roms yet...

**Reply by sgrockfort** [reply]
Hi @vossstef  ,

oooh, I can confirm that the cartridge freeze is triggered by pressing F11 
From this point of view, there is no reason to change anything. :-) 

Probably this is not the right place and hard to change, but just for your information I would like to share the following need:

Game cartridges work fine, but it seems that the most common fast loading cartridges do not work. I've also tried them with the default kernel and the 1541 default ROM... thanks Dolphin or Jiffy Dos, a quick load cartridge is basically unnecessary. 
I like using these cartridges because they can pause the CPU through the freeze button and provide the ability to strip the ram, change values/code, and return to the pre-freeze state. I still think it might be difficult to implement as some of these cartridges also include  extensions (RAM/mc).  also software emulators like vice have a hard time with this.  

I have tried Final Cartridge, Action-replay cartridge and Magic Formula. They crash by entering back the pre-freeze state.

If there is a way to do this in the future, that would be great.

Many thanks for your great work !!!!!!!!! 

cheers
Michael

**Reply by vossstef** [reply]
Hi @sgrockfort, 
as the C64Nano is based on the Mist'er core did you had a look in the forum how the guys there tackled the freezer topic ?
https://misterfpga.org/viewforum.php?f=12
https://misterfpga.org/viewtopic.php?t=7558
pls. let me know if you figure out if fully supported by Mist'er. Maybe i missed to port over something.

**Reply by sgrockfort** [reply]
HI @vossstef  ,

I tried the snappy rom which the guy said works fine with the c64 Mister-core. (https://misterfpga.org/viewtopic.php?t=7558)
I tried both available releases but unfortunately it didn't work.. I just got an empty black screen. Also I tried different combinations of kernel/1541 roms. 

Of course we don't know it the information from the post is reliable and I haven't a Mister to compare...

I will test it out some more, and search for more info to this topic in the mister forum the next days.. 

If I'll find something I'll let you know of course.

Thanks and nice weekend
Michael

**Reply by vossstef** [reply]
@sgrockfort it looks that once again the multicart crt load feature got broken...  ->black screen
will try to fix that first and then have a look into the freezer topic.

**Comment by vossstef** [comment]
@sgrockfort ,
i just did a 1.5.2 release that removes the black screen of snappy V5.34 when you load it.
You have to use a standard kernal. Looks that Dolphin is not compatible.
pls. try and let me know

Should snappy come up when you press F11 (RESTORE) or RESET via OSD ?

**Reply by vossstef** [reply]
@sgrockfort pls try together with the new .fs also the misterynano  that i added in the release ad i found out that there is a difference in btw companion and misterynano.

**Reply by sgrockfort** [reply]
Thank you @vossstef  , that's great news :-) :-) I will check it a bit later today and give you an update :-)

**Comment by sgrockfort** [comment]
Hello @vossstef  ,

I updated to 1.5.2 and also flashed the m0sdock with the firmware you uploaded to the release.. 
Also I chose CBM-DOS for the 1541.... For the C64-Kernal I tried two variants (no disk) and Kernal 901227-03

Final Cartridge,  ActionReplay and snappy Rom are starting now .. That's great :-)
Magic Formel V2 still doesn't start

About the freeze-Button: 

- with the snappy rom nothing happend when I press F11
- with Action-ReplayV6 it crashed (see attached image)

![1729335680106](https://github.com/user-attachments/assets/cd7a32c3-7453-4c2f-8efc-fad8fd0e9dfd)

Have a nice day
Michael

**Reply by vossstef** [reply]
@sgrockfort  thx!

need to dig into the F11 Button crash...

**Reply by sgrockfort** [reply]
@vossstef  thank **YOU** , for your effort. If I can help with testing, just let me know !

**Reply by vossstef** [reply]
@sgrockfort,
i got the freezer working ! Will release today after a lot of needed testing...

**Comment by vossstef** [comment]
@sgrockfort I created a 1.6.0 pre-release supporting the freezer function.
Pls. let me know... thx

**Reply by sgrockfort** [reply]
@vossstef  , perfect thank you... I'm going to test

**Reply by sgrockfort** [reply]
@vossstef  , I flashed the tang and m0sdock with the files from V1.6.0..
The CRTs are loading and the freeze button bring up the corresponding menu... I also managed to create a snapshot on disk. (with snappy rom CRT)

But now I have problems with accessing the drive... Sometimes I cannot load even the directory.  Even without any CRT and with standard ROMS. But not always.. After changing settings (removing CRT, change ROM, cold-reboot, unplug power) it works again. But not for a long time.  I couldn't find the systematic yet. Sometimes I also have just a black screen during startup (also without CRT)..  I will try a downgrade and test if the behavior still exists. Maybe it's also something wrong with my physical setup after flashing. I will send an update. unfortunately, probably I have a time not before wednesday.... 

Thank you so far..

Michael

**Reply by vossstef** [reply]
thanks for your help !
Just fixing a number to topics in the core to make it fully compatible with the companion FW.
Hope that those glitches you saw will go by that too.

**Reply by sgrockfort** [reply]
@vossstef  , I just made the following tests. Maybe the info is helpful 

first I only downgrade the M0SDock to fpga_companion with the binary from version 1.5.1 and keep tang on V 1.6.0
I had the same problems with accessing disk like mention above.

after that I also downgrade the tang to v 1.5.1.. 
After that I had no Problems with loading while 1541 and C64 are on standard and no Problems when both are on jiffy dos.
But with 1541 on dolpphin and C64 is on "no disk" (also dolphin) the loading also failed

I did all tests without any CRT to reduce complexity.

I am a bit short in time next days, but here and there I'd enjoy helpimg with testing.

**Reply by vossstef** [reply]
I just created a release v1.6.1

Unfortunately i found too many issues and therefore decided right away to release it. 
Propose to go for the FPGA-Companion FW + XML as i fixed also there several issues.  
CRT Multiload and Action Replay look to be working. Also the OSD Cartridge unload improves a lot the CRT use.

Pls let me knoe when you find time !

**Reply by sgrockfort** [reply]
Hello @vossstef  ,

I installed 1.6.1 as recommended ! The release looks much better to me :-)

- ActionReplay6 and snappy-rom are booting and the freeze menu also works
- created and restored a snapshot with snappy-rom
- snappy-rom fastloader works
- snappy-rom asm-monitor works
- the "detach CRT" menu is very helpful
- Jiffy (1541 & c64) works fine without CRT


it looks to me that snappy fully works - like the guy in the Mister-forum wrote !
great job !!!! :-)

in addition just as an Info:
- Action-Reply Fastloader not working (who knows if it works with Mister)
- dolphindos seems not to work anymore / even without any CRT

overall, as for me my needs are satisfied and I thank you for all your work!!!!
Anyways, if you need someone for testing - just let me know.. 

Michael

**Reply by vossstef** [reply]
@sgrockfort Thanks!

Let me check DolphinDOS again...

**Comment by vossstef** [comment]
@sgrockfort , please delete the c64.xml from the sdard for force the companion fw to fallback buildin configuration.

In my c64.xml the entries for c1541 reset in case of DOS rom change and Video PAL/NTSC mode change as still missing. my fault...

**Comment by vossstef** [comment]


**Reply by sgrockfort** [reply]
@vossstef  , i replaced the xml and it didn't help with the drive-access... I noticed that as soon as I change the 1541 rom it worked... So I reprogrammed the 2dosa_c file and now also dolphin also works.. Seems for some reason it was broken.. 
Sorry in case I produced confusion here. Now everything works fine. THX !!!

**Reply by vossstef** [reply]
@sgrockfort, 
if you attempt to flash the .FS without keeping the FPGA in reset by pressing S2 button during power-up and keeping for another ~5 sec you will very likely corrupt the additional stored images in FLASH causing cumbersome reprogram needs...

**Reply by sgrockfort** [reply]
I just noticed that always 1541-roms are damaged after flashing one... read somehwere that it's maybe due to the openfpga loader ubuntu package... (I took care of the S2-Button :-) ) ... However, I made one 128k file which contains all 1541-roms at the right offset and flashed them at once... Now all 1541 roms are working :-)

**Reply by vossstef** [reply]
@sgrockfort ,
le me check the flash blocksize (datasheet / openfpgaloader) if there is a fundamental problem in my flash sw production description...

I created a pre-release 1.6.2 to overcome the PAL/NTSC mode switch malfunction.  If you like have a try...

**Reply by sgrockfort** [reply]
@vossstef  imho there is no problem with the descripton. While usiing the windows gui tool there didn't appear such problems. 
I think it's about the particular package from the linux fpgaloader I used. The workaround I made together with @geniac works very well and is very comfortable. :-)

dd if=/dev/zero of=combined.bin bs=1 count=131072
dd if=dosa.bin of=combined.bin bs=1 conv=notrunc
dd if=cbmdos.bin of=combined.bin bs=1 seek=49152 conv=notrunc
dd if=c1541.bin of=combined.bin bs=1 seek=81920 conv=notrunc
dd if=jiffy.bin of=combined.bin bs=1 seek=114688 conv=notrunc

openFPGALoader -b tangnano20k --external-flash -o 0x200000 combined.bin

Thank for the new release. I will test it :)

**Comment by sgrockfort** [comment]
@vossstef  I tested the 1.6.2 and switch about 10 times (PAL<>NTSC)... 
It works perfect... Also I couldn't find any (new) side effects...  :+1:

**Reply by vossstef** [reply]
Big thanks for testing !

**Reply by sgrockfort** [reply]
@vossstef  , just noticed a bug in 1.6.2 .... everything is a bit slower/lagging... you will notice it just in a game... Sound and movement are lagging.... This happens in PAL & NTSC mode... To ensure no other problems,  I downgrade to 1.6.1 and the problem is gone... I tested it twice... very sorry that I didn't saw it yesterday !

**Reply by vossstef** [reply]
@sgrockfort i will look into that tomorrow. 
Could you pls open an issue for that and write the game you tested that laggs? thx

**Reply by sgrockfort** [reply]
done.... thx
